### ISSUE TITLE
Import code from the host project

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#8473]
Sometimes a Dagger module needs to import code from the project it's embedded in. For example, in our own CI we use a `distconsts` package to share constants across the Dagger codebase, and the pipelines shipping it. Things like the default path of the engine's state directory; etc.

There's no officially recommended way to do this. What Dagger does for its own CI, is not mentioned in the docs, and we don't know whether it's a good pattern, because we haven't discussed it. So let's discuss it :)

**Comment by Excoriate** [comment]
In a polyglot monorepo setup, how would this work? Should it be aligned with the Dagger module's SDK, meaning it can't import a pip package in a Dagger module built using the Go SDK?
Overall, it's a great idea.

**Comment by pjmagee** [comment]
> In a polyglot monorepo setup, how would this work? Should it be aligned with the Dagger module's SDK, meaning it can't import a pip package in a Dagger module built using the Go SDK? Overall, it's a great idea.

Multiple dagger sdks in one repo... is that an option? maybe that's not sane though. But, if a team is working in a polyglot way, if dagger does support that language, why not have multiple dagger sdks? - Would be cool to rollup into something like `dagger functions` - but it knows to pull in like several dagger subfolders where each sdk is a diff language. Maybe this opens a can of worms though, like complexity/confusion and hard to maintain?

**Comment by helderco** [comment]
One solution that works today is to leverage module dependencies, as described in:
- https://github.com/dagger/dagger/issues/8496

But I think allowing include/exclude patterns in a `dagger.json` to reach the context directory, would provide more fine-grained control, and it would be easier to manage:
- https://github.com/dagger/dagger/issues/8531

**Comment by shykes** [comment]
@Excoriate @pjmagee yes, for polyglot monorepos, it's ideal to have different modules in different parts of the repo, each with its own SDK.

But sometimes one of those modules will need to import code from the application module within the same language silo. That's what this issue is about.

Using native imports as an alternative to multiple modules is definitely not recommended, edit in a polyglot repo.

**Comment by helderco** [comment]
> But I think allowing include/exclude patterns in a `dagger.json` to reach the context directory, would provide more fine-grained control, and it would be easier to manage:
> 
> * [Allow `dagger.json` include/exclude patterns to match context directory #8531](https://github.com/dagger/dagger/issues/8531)

Correction, it's actually doable with `include`/`exclude` patterns to import code from outside the module, but could be better with:
- https://github.com/dagger/dagger/issues/8548
- https://github.com/dagger/dagger/issues/8547
